sdk/include/sdk/Logger.h

```diff
@@ -17,71 +17,129 @@
  *
  */
-class ConsoleLogger {
+class ILogger {
 public:
-    void info(const std::string& msg) const {
-        fmt::print(fmt::fg(fmt::color::white), "{}, INFO  : {}\n", std::chrono::system_clock::now(),
-                   msg);
-        std::fflush(stdout);
-    }
+    ILogger()          = default;
+    virtual ~ILogger() = default;
 
-    void warn(const std::string& msg) const {
-        fmt::print(fmt::fg(fmt::color::yellow), "{}, WARN  : {}\n",
-                   std::chrono::system_clock::now(), msg);
-        std::fflush(stdout);
-    }
+    ILogger(const ILogger&)            = delete;
+    ILogger(ILogger&&)                 = delete;
+    ILogger& operator=(const ILogger&) = delete;
+    ILogger& operator=(ILogger&&)      = delete;
 
-    void error(const std::string& msg) const {
-        fmt::print(fmt::fg(fmt::color::red), "{}, ERROR : {}\n", std::chrono::system_clock::now(),
-                   msg);
-        std::fflush(stdout);
-    }
+    /**
+     * @brief Log a message with info level.
+     *
+     * @param msg The message to log.
+     */
+    virtual void info(const std::string& msg) = 0;
 
-    void debug(const std::string& msg) const {
-        fmt::print(fmt::fg(fmt::color::brown), "{}, DEBUG : {}\n", std::chrono::system_clock::now(),
-                   msg);
-        std::fflush(stdout);
-    }
+    /**
+     * @brief Log a message with warn level.
+     *
+     * @param msg The message to log.
+     */
+    virtual void warn(const std::string& msg) = 0;
+
+    /**
+     * @brief Log a message with error level.
+     *
+     * @param msg The message to log.
+     */
+    virtual void error(const std::string& msg) = 0;
+
+    /**
+     * @brief Log a message with debug level.
+     *
+     * @param msg The message to log.
+     */
+    virtual void debug(const std::string& msg) = 0;
 };
 
 /**
- * @brief Logger template implementation which will delegate the calls to the
- * implementing class.
+ * @brief Allows logging of messages with different log levels.
  *
- * @tparam TImpl  The class implementing the logger functionality.
+ * @details Handles string formatting of the log message and its arguments
+ *          then forwards the formatted message to the real logger
+ *          implementation. (Bridge pattern)
  */
-template <typename TImpl> class Logger {
+class Logger {
 public:
-    void info(const std::string& msg) const { m_impl.info(msg); }
-    void warn(const std::string& msg) const { m_impl.warn(msg); }
-    void error(const std::string& msg) const { m_impl.error(msg); }
-    void debug(const std::string& msg) const { m_impl.debug(msg); }
+    Logger();
+
+    /**
+     * @brief Log a message with info level.
+     *
+     * @tparam T    Type of the format arguments.
+     * @param msg   The format message.
+     * @param args  The format arguments.
+     */
+    template <typename... T> void info(const std::string& msg, const T&... args) {
+        m_impl->info(fmt::format(msg, args...));
+    }
```

Comment:
Basically fine. 
But I'm wondering, if we shouldn't use the streaming approach (like cout/cerr) for this instead. That would allow for reusing existing ostream ops of types and classes ..

Reply:
Personally, I am not a big fan of streaming operator for logging. We can provide streaming operators in addition to regular log statements. Whats to note is that fmtlibs speed is better than ostream and compile size is also lower than iostream. See https://github.com/fmtlib/fmt